keps/0007-command-asserts.md

```diff
@@ -0,0 +1,103 @@
+- Keep eventual consistency when asserting commands
+
+### Non-Goals
+
+- Introduce "flows" to determine when to run commands. I.e. to run commands once before or after asserting resources
+
+## Proposal
+
+Introduce commands as part of `TestAssert`. These commands are run at a regular interval like resource assertions until they either succeed or a timeout of the assert is reached. If any of the commands fail, the assert is considered to have failed. A command's return value determines if a command failed or not.
+
+As a start, complex commands that parse output can be described using shell environments. At a later stage we can implement additional tools to simplify common tasks. E.g. to search with a regular expression in command output.
+
+### User Stories
+
+#### User Story 1
+
+The JVM options of a KUDO operator can be tuned using parameters. On deployment, the JVM options are logged in the pods running the application provided by the operator. To assert that a parameter change resulted in changed JVM options we need to run `kubectl logs pod-name` and search the logs for a specific string once the operator has been deployed.
+
+#### User Story 2
+
+An operator can be configured to encrypt it's API. To assert that the API endpoints are indeed encrypted we can run a specific `curl` command and check that it returns successfully once the operator has been deployed.
+
+### Implementation Details
+
+Command assertions use most fields of the existing commands for `TestStep` and follow their behavior. The `ignoreFailure`, `background` and `timeout` fields are removed as they are not needed here.
+
+An example that parses the logs of a pod:
+
+```yaml
+apiVersion: kuttl.dev/v1beta1
+type: TestAssert
+commands:
+  - command: ./parse-logs.sh name-of-pod-0 string-to-search-for
+    namespaced: true
+```
+
+with `parse-logs.sh`:
+
+```bash
+#!/bin/bash
+
+POD=$1
+SEARCH=$2
+# $3 is "--namespace"
+NAMESPACE=$4
+
+kubectl logs -n ${NAMESPACE} ${POD} | grep ${SEARCH}
+```
+
+### Risks and Mitigations
+
+Some commands might be expensive to run and it would be better to run them once instead of running them repeatedly. This can be worked around by running commands as part of a test step instead. Commands in a test step run once.
+
+There could be subtle race conditions when running commands that have to run after a `TestStep` has been applied. The eventual consistency of Kubernetes might result in a command running in an environment where the changes of the `TestStep` haven't been applied yet. E.g., consider a `TestStep` that changes a `Deployment` and a command assert that examines the log of a `Pod` run by the `Deployment`. The change of the `Deployment` will restart pods, but as the restart is not immediate, the command might be run with the old pod.
+As a workaround, users can create an assert that doesn't have corresponding `TestStep`. This will then get asserted after the prior assert succeeded.
```

Comment:
yes!   this is what I meant in my last comment... 00-assert with declaratives and 01-assert with commands.

however... we should consider the following:

```
apiVersion: kuttl.dev/v1beta1
type: TestAssert
commands:
  before:
    - command: ./parse-logs.sh name-of-pod-0 string-to-search-for
  after:
    - command: ./parse-logs.sh name-of-pod-0 string-to-search-for
```

where we can find better words for `before` and `after`... and they mean before declarative asserts and after.

We could also just always have them after (opposite of previous earlier advice).   If they are always after we know that a transition can be confirmed before the command assert.  This is likely a common scenario IMO.

We should have these captured as alt proposals.   I'm leaning on the last recommendation now tho... cycling through all commands after declarative.

Reply:
This is currently a non-goal, though it makes sense to add this later.